File: lib/modules/taxi/utils/route_destination_trim.dart
import 'package:latlong2/latlong.dart';

import 'taxi_distance_calculator.dart';

/// عند الاقتراب من الوجهة على الجانب الآخر من الشارع قد يُطيل الموجّه المسار
/// لاستدارة قانونية. نُنهي المسار عند أول نقطة ضمن هذه العتبة.
const double routeNearDestinationTrimMeters = 200;

class RouteDestinationTrimResult {
  final List<LatLng> points;
  final double distanceMeters;
  final int? durationSeconds;
  final bool trimmed;

  const RouteDestinationTrimResult({
    required this.points,
    required this.distanceMeters,
    this.durationSeconds,
    this.trimmed = false,
  });
}

RouteDestinationTrimResult trimRouteNearDestination({
  required List<LatLng> points,
  required LatLng destination,
  double thresholdMeters = routeNearDestinationTrimMeters,
  double? distanceMeters,
  int? durationSeconds,
}) {
  if (points.length < 2) {
    final meters = distanceMeters ?? 0;
    return RouteDestinationTrimResult(
      points: points,
      distanceMeters: meters,
      durationSeconds: durationSeconds,
    );
  }

  final originalDistance = (distanceMeters != null && distanceMeters > 0)
      ? distanceMeters
      : _distanceAlongPolyline(points);

  final near = _findFirstNearPointOnRoute(points, destination, thresholdMeters);
  if (near == null) {
    return RouteDestinationTrimResult(
      points: points,
      distanceMeters: originalDistance,
      durationSeconds: durationSeconds,
    );
  }

  final tailDistance = _distanceAlongPolyline(points.sublist(near.cutIndex + 1));
  if (tailDistance < 30) {
    return RouteDestinationTrimResult(
      points: points,
      distanceMeters: originalDistance,
      durationSeconds: durationSeconds,
    );
  }

  final trimmedPoints = <LatLng>[
    ...points.sublist(0, near.cutIndex + 1),
    near.cutPoint,
    destination,
  ];
  final trimmedDistance = near.distanceAlongMeters + near.remainderMeters;

  int? trimmedDuration = durationSeconds;
  if (durationSeconds != null && originalDistance > 0) {
    final ratio = (trimmedDistance / originalDistance).clamp(0.0, 1.0);
    trimmedDuration = (durationSeconds * ratio).round().clamp(30, durationSeconds);
  }

  return RouteDestinationTrimResult(
    points: trimmedPoints,
    distanceMeters: trimmedDistance,
    durationSeconds: trimmedDuration,
    trimmed: true,
  );
}

class _NearCut {
  final LatLng cutPoint;
  final double distanceAlongMeters;
  final double remainderMeters;
  final int cutIndex;

  const _NearCut({
    required this.cutPoint,
    required this.distanceAlongMeters,
    required this.remainderMeters,
    required this.cutIndex,
  });
}

_NearCut? _findFirstNearPointOnRoute(
  List<LatLng> points,
  LatLng destination,
  double thresholdMeters,
) {
  var accumulatedMeters = 0.0;

  for (var i = 0; i < points.length; i++) {
    final point = points[i];
    final pointDistance = _distanceMeters(point, destination);
    if (pointDistance <= thresholdMeters) {
      return _NearCut(
        cutPoint: point,
        distanceAlongMeters: accumulatedMeters,
        remainderMeters: pointDistance,
        cutIndex: i,
      );
    }

    if (i >= points.length - 1) continue;

    final next = points[i + 1];
    final segmentMeters = _distanceMeters(point, next);
    if (segmentMeters <= 0) continue;

    final steps = (segmentMeters / 8).ceil().clamp(12, 80);
    for (var step = 1; step <= steps; step++) {
      final t = step / steps;
      final sample = LatLng(
        point.latitude + (next.latitude - point.latitude) * t,
        point.longitude + (next.longitude - point.longitude) * t,
      );
      final sampleDistance = _distanceMeters(sample, destination);
      if (sampleDistance <= thresholdMeters) {
        return _NearCut(
          cutPoint: sample,
          distanceAlongMeters: accumulatedMeters + segmentMeters * t,
          remainderMeters: sampleDistance,
          cutIndex: i,
        );
      }
    }

    accumulatedMeters += segmentMeters;
  }

  return null;
}

double _distanceAlongPolyline(List<LatLng> points) {
  if (points.length < 2) return 0;
  var total = 0.0;
  for (var i = 1; i < points.length; i++) {
    total += _distanceMeters(points[i - 1], points[i]);
  }
  return total;
}

double _distanceMeters(LatLng a, LatLng b) {
  return TaxiDistanceCalculator.calculateDistance(
        a.latitude,
        a.longitude,
        b.latitude,
        b.longitude,
      ) *
      1000;
}

File: backend/lib/route_destination_trim.js
/**
 * عند الاقتراب من الوجهة (على الجانب الآخر من الشارع) قد يُطيل Mapbox المسار
 * لاستدارة قانونية. نُنهي المسار عند أول نقطة ضمن عتبة قريبة ولا نحسب الاستدارة.
 */

const NEAR_DESTINATION_THRESHOLD_METERS = Number.parseInt(
  process.env.ROUTE_NEAR_DESTINATION_METERS || '200',
  10
);

function haversineDistanceMeters(origin, destination) {
  const earthRadiusMeters = 6371000;
  const toRadians = (value) => (value * Math.PI) / 180;
  const lat1 = toRadians(origin.latitude);
  const lat2 = toRadians(destination.latitude);
  const deltaLat = toRadians(destination.latitude - origin.latitude);
  const deltaLng = toRadians(destination.longitude - origin.longitude);
  const a =
    Math.sin(deltaLat / 2) * Math.sin(deltaLat / 2) +
    Math.cos(lat1) *
      Math.cos(lat2) *
      Math.sin(deltaLng / 2) *
      Math.sin(deltaLng / 2);
  return earthRadiusMeters * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function lerpPoint(a, b, t) {
  return {
    latitude: a.latitude + (b.latitude - a.latitude) * t,
    longitude: a.longitude + (b.longitude - a.longitude) * t,
  };
}

function distanceAlongPolyline(points) {
  if (!Array.isArray(points) || points.length < 2) return 0;
  let total = 0;
  for (let i = 1; i < points.length; i += 1) {
    total += haversineDistanceMeters(points[i - 1], points[i]);
  }
  return total;
}

function findFirstNearPointOnRoute(points, destination, thresholdMeters) {
  if (!Array.isArray(points) || points.length === 0) return null;

  let accumulatedMeters = 0;
  const dest = {
    latitude: Number(destination.latitude),
    longitude: Number(destination.longitude),
  };

  for (let i = 0; i < points.length; i += 1) {
    const point = points[i];
    const pointDistance = haversineDistanceMeters(point, dest);
    if (pointDistance <= thresholdMeters) {
      return {
        cutPoint: point,
        distanceAlongMeters: accumulatedMeters,
        remainderMeters: pointDistance,
        cutIndex: i,
      };
    }

    if (i >= points.length - 1) continue;

    const next = points[i + 1];
    const segmentMeters = haversineDistanceMeters(point, next);
    if (segmentMeters <= 0) continue;

    const steps = Math.max(12, Math.ceil(segmentMeters / 8));
    for (let step = 1; step <= steps; step += 1) {
      const t = step / steps;
      const sample = lerpPoint(point, next, t);
      const sampleDistance = haversineDistanceMeters(sample, dest);
      if (sampleDistance <= thresholdMeters) {
        return {
          cutPoint: sample,
          distanceAlongMeters: accumulatedMeters + segmentMeters * t,
          remainderMeters: sampleDistance,
          cutIndex: i,
        };
      }
    }

    accumulatedMeters += segmentMeters;
  }

  return null;
}

function trimRouteNearDestination(
  points,
  destination,
  {
    thresholdMeters = NEAR_DESTINATION_THRESHOLD_METERS,
    durationSeconds = null,
    distanceMeters = null,
  } = {}
) {
  const normalizedPoints = Array.isArray(points)
    ? points
        .map((entry) => ({
          latitude: Number(entry.latitude),
          longitude: Number(entry.longitude),
        }))
        .filter(
          (entry) =>
            Number.isFinite(entry.latitude) && Number.isFinite(entry.longitude)
        )
    : [];

  const dest = {
    latitude: Number(destination?.latitude),
    longitude: Number(destination?.longitude),
  };

  const originalDistance =
    Number.isFinite(distanceMeters) && distanceMeters > 0
      ? distanceMeters
      : distanceAlongPolyline(normalizedPoints);
  const originalDuration =
    Number.isFinite(durationSeconds) && durationSeconds > 0
      ? durationSeconds
      : null;

  if (
    normalizedPoints.length < 2 ||
    !Number.isFinite(dest.latitude) ||
    !Number.isFinite(dest.longitude)
  ) {
    return {
      points: normalizedPoints,
      distanceMeters: originalDistance,
      durationSeconds: originalDuration,
      trimmed: false,
    };
  }

  const near = findFirstNearPointOnRoute(
    normalizedPoints,
    dest,
    thresholdMeters
  );
  if (!near) {
    return {
      points: normalizedPoints,
      distanceMeters: originalDistance,
      durationSeconds: originalDuration,
      trimmed: false,
    };
  }

  const tailDistance = distanceAlongPolyline(
    normalizedPoints.slice(near.cutIndex + 1)
  );
  if (tailDistance < 30) {
    return {
      points: normalizedPoints,
      distanceMeters: originalDistance,
      durationSeconds: originalDuration,
      trimmed: false,
    };
  }

  const head = normalizedPoints.slice(0, near.cutIndex + 1);
  const trimmedPoints = [...head, near.cutPoint, dest];
  const trimmedDistance = near.distanceAlongMeters + near.remainderMeters;

  let trimmedDuration = originalDuration;
  if (originalDuration != null && originalDistance > 0) {
    const ratio = Math.min(1, trimmedDistance / originalDistance);
    trimmedDuration = Math.max(30, Math.round(originalDuration * ratio));
  }

  return {
    points: trimmedPoints,
    distanceMeters: trimmedDistance,
    durationSeconds: trimmedDuration,
    trimmed: true,
  };
}

module.exports = {
  NEAR_DESTINATION_THRESHOLD_METERS,
  haversineDistanceMeters,
  trimRouteNearDestination,
};
